source/slang/slang-check-decl.cpp

```diff
@@ -775,6 +775,24 @@ namespace Slang
         return true;
     }
 
+    void SemanticsVisitor::_validateCircularVarDefinition(VarDeclBase* varDecl)
+    {
+        // The easiest way to test if the declaration is circular is to
+        // validate it as a constant.
+        //
+        // TODO: The logic here will only apply for `static const` declarations
```

Comment:
Whilst this does root out the more subtle scenarios - this problem was originally identified with the most straightforward 

```
static const SomeType kThing = kThing;
```

Can't we also have an additional simple test for this case and it doesn't have the requirement of only being constant foldable (and therefore currently int)?

Reply:
To be clear, there are two main places in the compiler where this kind of circularity can cause a crash:

1. In the front-end, when folding `int` constants

2. In the AST->IR lowering step, when dealing with `static const`s of all types

This PR is intentionally just addressing (1) by adding protection to the front-end constant-folding path (and then hijacking that logic to also provide a guaranteed error message even if a constant is never referenced).

I'd like to attack (2) with a second PR, and that PR would catch the case you show as well as others.

The current PR isn't meant to dismiss the importance of non-`int` types; it is only as narrowly scoped as it is because the constant folding logic that we are trying to protect is only applicable to integers.